### PR TITLE
Navigation decoupling using events

### PR DESCRIPTION
**Pull request by OceanWolf**
Have only made this work on TkAgg, will convert for GTK3 later.  In fixing `ConfigureSubplotsTk`, I have noticed a marked decrease in dialogue reopening time :), I guess the "figure" creation slowed it down.

I have noticed problems with the home/back/forward buttons on testing, though I guess this bug already existed, so I will look into it and fix it with a separate PR later.

### CODE REVIEW

**Comment by fariza** [comment]
I will comment on specific lines as I understand the changes but the one thing that is clear.

Please split this in two PRs.
- One for the tool instances. This I'm convinced.
- Other for the Google handling and moving the add tool to the toolbar. This I'm not convinced

**Comment by fariza** [comment]
Don't worry for the GTK changes. When we finish all the changes I can take care of that

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -3300,9 +3294,10 @@ def unregister(self, name):
         """
 
         if self._toggled == name:
-            self._handle_toggle(name, from_toolbar=False)
-        if name in self._instances:
-            del self._instances[name]
+            self._handle_toggle(name)
+        if name in self._tools:
```
Comment (fariza) [inline lib/matplotlib/backend_bases.py:3298]: In what case, name is not present in `self._tools`?

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -3361,24 +3355,12 @@ def add_tool(self, name, tool, position=None):
...
-                fname = os.path.join(basedir, tool_cls.image)
-            else:
-                fname = None
-            toggle = issubclass(tool_cls, tools.ToolToggleBase)
-            self.toolbar._add_toolitem(name,
-                                       tool_cls.description,
-                                       fname,
-                                       position,
-                                       toggle)
+            self.toolbar.add_tool(name, self._tools[name], position)
```
Comment (fariza) [inline lib/matplotlib/backend_bases.py:3363]: Do we want users adding tools to the `toolbar` directly? I think the way is through `navigation`, this method should be `self.toolbar._add_tool(name, self._tools[name], position)

**Comment by OceanWolf** [comment]
Hmm, I thought I may have done too much for one pull request, but it seemed like one would go in as a very small request, the other much larger, but I can try and separate them out and see, in which case it will become a part 2 and a part 3.

Do you know of a way to do that on github?  If not, would it make it easier for you to wait until I have split the PRs before you start commenting?

What do you mean by `Google handling`?

**Comment by fariza** [comment]
Sorry, I don't know.
You can leave this PR open, and do the other in another branch, so it's easier for you to follow your changes.

**Comment by fariza** [comment]
Please describe the problem with the `home/back/forward` buttons. I haven't noticed anything.

**Comment by OceanWolf** [comment]
Problem with home/back/forward:
it seems fine when I switched back, so it comes from me.  I shall investigate.  (it seems to forget the history).

PR:
If I understand correctly, I should revert all changes to this branch, except for the minimum of code required to keep tool instances alive, i.e. `_instances` -> `tools`.
I do that by taking a branch of this as is.  Then switching back to this branch and taking the stuff out.

I will probably do this tomorrow as it gets late here.

**Comment by fariza** [comment]
What I would do (I'm far from expert) I would create a new branch based on the original code. And port only the instances relate code. This way you keep this branch as reference. But do it as you like, at the end is the same

**Comment by OceanWolf** [comment]
Reading up on the subject, I think I will perform a rebase as it seems cleaner.

Thinking about the code, I can actually see three PRs here, dividing the tool instances into two parts.  Separating out the code should help you to see the logic behind the various bits of code.

I will let you know when it has become "safe" the code again ;).

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (OceanWolf) [inline lib/matplotlib/backend_bases.py:3363]: I disagree, though not an issue for here after my rebase splitting.

When I put this patch in (a different PR) I shall put it in as private and as and when I have figured out my preferred way of adding tools (on paper it looks almost done), I can make it public again (for you to comment on), if it still feels right.

**Comment by fariza** [comment]
@OceanWolf 
I'm going to be out of reach  for the next two weeks (11 to 22), just in case you wonder why I don't answer.

**Comment by OceanWolf** [comment]
@fariza No problem, my laptop died on me on Monday (general system board failure :'(), so trying to sort something out.  But yes, the reason for my silence so far.

Enjoy your holiday!

On Wednesday, 6 August 2014, 15:02, Federico Ariza notifications@github.com wrote:

@OceanWolf 
I'm going to be out of reach  for the next two weeks (11 to 22), just in case you wonder why I don't answer.

**Comment by fariza** [comment]
@OceanWolf I'm back, did you have time to advance on the split of the PR?

**Comment by OceanWolf** [comment]
No, sorry, laptop still fried.  Slowly upgrading this desktop to a point where I can work on it.  Hopefully only a few days more.

As I remember, when I split it, the first patch contained around 20 lines of additions, and around 80 deletions, or something like that anyway.

**Comment by fariza** [comment]
@OceanWolf 
I just removed Persistent Tools before going any further you have to rebase your code

**Pull request closed without merging**